core/src/main/java/org/infinispan/remoting/inboundhandler/NonTotalOrderPerCacheInboundInvocationHandler.java

```diff
@@ -27,15 +57,44 @@ public void handle(CacheRpcCommand command, Reply reply, DeliverOrder order) {
       }
       try {
          final int commandTopologyId = extractCommandTopologyId(command);
-         final boolean waitTransactionalData = command.getCommandId() != StateRequestCommand.COMMAND_ID;
          final boolean onExecutorService = executeOnExecutorService(order, command);
-         handleRunnable(createDefaultRunnable(command, reply, commandTopologyId, waitTransactionalData, onExecutorService), onExecutorService);
+         final BlockingRunnable runnable;
+
+         switch (command.getCommandId()) {
+            case SingleRpcCommand.COMMAND_ID:
+               //lock ownership needs the first topology. we need to check it before invoking the getLockPromise()
+               if (isCommandSentBeforeFirstTopology(commandTopologyId)) {
+                  reply.reply(CacheNotFoundResponse.INSTANCE);
+                  return;
+               }
+               runnable = createLockAwareRunnable(command, reply, commandTopologyId, true, onExecutorService,
+                                                  getLockPromise((SingleRpcCommand) command));
```

Comment:
By async commands I meant the write commands and the prepare command in DIST_ASYNC/REPL_ASYNC, which are executed on JGroups' regular thread pool.
You're right that the remote get doesn't acquire the locks itself, but it does block the OOB thread if `acquireRemoteLock == true`, so maybe it should be covered by this as well?